File: bindings/python/samples/sample_support.py
import socket as _socket
import threading
import time

import zlink


_ALLOCATED_TCP_PORTS = set()


def tcp_endpoint():
    while True:
        sock = _socket.socket(_socket.AF_INET, _socket.SOCK_STREAM)
        sock.bind(("127.0.0.1", 0))
        port = sock.getsockname()[1]
        sock.close()
        if port not in _ALLOCATED_TCP_PORTS:
            _ALLOCATED_TCP_PORTS.add(port)
            return port, f"tcp://127.0.0.1:{port}"


def _poll_monitor_event(monitor, timeout_ms):
    with zlink.create_poller() as poller:
        poller.add_monitor(monitor, zlink.PollEventFlag.POLLIN, 0)
        events = zlink.create_poll_events(1)
        try:
            if poller.wait(events, timeout_ms):
                return monitor.recv(flags=zlink.RecvFlags.DONT_WAIT)
            return None
        finally:
            poller.remove_monitor(monitor)


def wait_connected(*monitors, timeout_ms=5000):
    pending = list(monitors)
    deadline = time.monotonic() + (timeout_ms / 1000.0)

    while pending:
        remaining_ms = int((deadline - time.monotonic()) * 1000)
        if remaining_ms <= 0:
            raise TimeoutError("connection handshake did not complete")

        next_pending = []
        for monitor in pending:
            event = _poll_monitor_event(monitor, remaining_ms)
            if event is None:
                next_pending.append(monitor)
                continue
            if not (int(event.event) & int(zlink.MonitorEventMask.CONNECTION_READY)):
                next_pending.append(monitor)
        pending = next_pending


def wait_socket_monitor_event(monitor, expected_event, timeout_ms=5000):
    deadline = time.monotonic() + (timeout_ms / 1000.0)
    expected_value = int(expected_event)

    while True:
        remaining_ms = int((deadline - time.monotonic()) * 1000)
        if remaining_ms <= 0:
            raise TimeoutError("socket monitor did not produce the expected event")
        event = _poll_monitor_event(monitor, remaining_ms)
        if event is None:
            continue
        if int(event.event) & expected_value:
            return event


def wait_until(predicate, timeout_ms=5000, description="condition"):
    deadline = time.monotonic() + (timeout_ms / 1000.0)
    waiter = threading.Event()
    while time.monotonic() < deadline:
        if predicate():
            return
        waiter.wait(0.001)
    raise TimeoutError(f"timed out waiting for {description}")


def submit_request_op(op, *, timeout=2, description="request"):
    try:
        messages = op.timeout(timeout).submit_sync()
    except zlink.RequestError as error:
        raise RuntimeError(f"{description} failed: {error.result!r}") from error
    try:
        return [message.to_bytes() for message in messages]
    finally:
        for message in messages:
            message.close()

File: bindings/rust/samples/sample_support.rs
#![allow(dead_code)]

use std::future::Future;
use std::pin::pin;
use std::sync::Arc;
use std::task::{Context, Poll, Wake, Waker};
use std::thread::{self, Thread};
use std::time::{Duration, Instant};

use zlink::{POLLIN, PollEvent, Poller, RecvFlags, SocketMonitor};

struct ThreadWake(Thread);

impl Wake for ThreadWake {
    fn wake(self: Arc<Self>) {
        self.0.unpark();
    }

    fn wake_by_ref(self: &Arc<Self>) {
        self.0.unpark();
    }
}

pub(crate) fn block_on<F: Future>(future: F) -> F::Output {
    let waker = Waker::from(Arc::new(ThreadWake(thread::current())));
    let mut context = Context::from_waker(&waker);
    let mut future = pin!(future);
    loop {
        match future.as_mut().poll(&mut context) {
            Poll::Ready(output) => return output,
            Poll::Pending => thread::park(),
        }
    }
}

pub fn tcp_endpoint() -> String {
    let listener = std::net::TcpListener::bind("127.0.0.1:0").unwrap();
    let port = listener.local_addr().unwrap().port();
    drop(listener);
    format!("tcp://127.0.0.1:{}", port)
}

pub fn wait_connected(monitors: &[&SocketMonitor]) {
    let poller = Poller::new().expect("monitor poller creation failed");
    let mut ready = vec![false; monitors.len()];
    for (slot, monitor) in monitors.iter().enumerate() {
        poller
            .add_monitor(monitor, POLLIN, slot)
            .expect("monitor poller registration failed");
    }

    let mut events = vec![PollEvent::default(); monitors.len().max(1)];
    while ready.iter().any(|item| !item) {
        let count = poller.wait(&mut events, -1).expect("monitor poll failed");
        for poll_event in events.iter().take(count) {
            let monitor = monitors[poll_event.slot];
            while let Some(event) = monitor
                .recv_with_flags(RecvFlags::DONT_WAIT)
                .expect("monitor recv failed")
            {
                if event.is_connection_ready()
                    || monitor.status().expect("monitor status failed").is_ready()
                {
                    ready[poll_event.slot] = true;
                }
            }
        }
    }
}

pub fn wait_stream_connected(monitor: &SocketMonitor) {
    let poller = Poller::new().expect("monitor poller creation failed");
    poller
        .add_monitor(monitor, POLLIN, 0)
        .expect("monitor poller registration failed");
    let mut events = [PollEvent::default()];
    loop {
        poller.wait(&mut events, -1).expect("monitor poll failed");
        while let Some(event) = monitor
            .recv_with_flags(RecvFlags::DONT_WAIT)
            .expect("monitor recv failed")
        {
            if event.is_accepted()
                || event.is_connection_ready()
                || monitor.status().expect("monitor status failed").is_ready()
            {
                return;
            }
        }
    }
}

pub fn wait_until<F>(mut predicate: F, timeout: Duration, description: &str)
where
    F: FnMut() -> bool,
{
    let deadline = Instant::now() + timeout;
    while Instant::now() < deadline {
        if predicate() {
            return;
        }
        std::thread::yield_now();
    }
    panic!("timed out waiting for {description}");
}
